File: Python/test/UnionFindSet.py
"""
并查集的定义与常用操作
"""

class UnionFindSet(object):

    def __init__(self, data_list):
        # 存储节点-->代表点的映射关系
        self.parent_dict = {}
        # 只有一个节点是某个集合的代表点时，才存到字典中
        self.size_dict = {}

        for node in data_list:
            self.parent_dict[node] = node  # 初始时，每个节点指向自己
            self.size_dict[node] = 1  # 每个集合一开始就1个节点

    def find_parent(self, node):
        """
        给定节点node，查找节点的代表点并返回
        优化：顺便把沿途所有节点的代表点直接改成最高的代表点
        """
        path = []  # 用于记录沿途的结点
        while node != self.parent_dict[node]:
            path.append(node)
            node = self.parent_dict[node]

        # 此时node来到了代表点
        while len(path) > 0:
            self.parent_dict[path.pop()] = node
        return node

    def is_same_set(self, node_a, node_b):
        """
        查询node_a和node_b是否在同一个集合里
        """
        return self.find_parent(node_a) == self.find_parent(node_b)

    def union(self, node_a, node_b):
        """
        将两个集合合并在一起
        """
        if node_a is None or node_b is None:
            return
        a_head = self.find_parent(node_a)
        b_head = self.find_parent(node_b)

        if a_head != b_head:  # 只在两者不在一个集合时进行下述操作
            a_set_size = self.size_dict[a_head]
            b_set_size = self.size_dict[b_head]
            # 优化：求集合大小，把小的挂在大的集合之下，减小遍历链条的长度
            if a_set_size >= b_set_size:
                self.parent_dict[b_head] = a_head
                self.size_dict[a_head] = a_set_size + b_set_size
                self.size_dict.pop(b_head, 0)  # 删除设定一个默认值，防止删掉不存在的key而报错
            else:
                self.parent_dict[a_head] = b_head
                self.size_dict[b_head] = a_set_size + b_set_size
                self.size_dict.pop(a_head, 0)

File: Java/src/structure/UnionFindSet.java
package structure;

import java.util.HashMap;
import java.util.List;
import java.util.Stack;

/**
 * 并查集
 */
public class UnionFindSet<V> {
    // 节点值-节点对象的映射关系，一经初始化就不会改变
    public HashMap<V, MyNode<V>> nodes;
    // 记录节点和父母，注意里面存的不一定就是最高的祖先
    public HashMap<MyNode<V>, MyNode<V>> parentMap;
    // 记录每个独立集合的节点数目,只有是祖先（最高的父母节点）的那些节点才会出现在这里
    public HashMap<MyNode<V>, Integer> sizeMap;

    public UnionFindSet(List<V> values) {
        // 初始化时，所有节点自成一个集合
        for (V val : values) {
            MyNode<V> node = new MyNode<>(val);
            nodes.put(val, node);
            parentMap.put(node, node);
            sizeMap.put(node, 1);
        }
    }

    // 查看两节点是否属于同一集合
    public boolean isSameSet(V a, V b) {
        if (!nodes.containsKey(a) || !nodes.containsKey(b)) {
            return false;
        }
        return findParent(nodes.get(a)) == findParent(nodes.get(b));
    }

    // 合并两个节点所属的集合
    public void union(V a, V b) {
        if (!nodes.containsKey(a) || !nodes.containsKey(b)) {
            return;
        }
        MyNode<V> aParent = findParent(nodes.get(a));
        MyNode<V> bParent = findParent(nodes.get(b));
        if (aParent != bParent) {
            int aSize = sizeMap.get(aParent);
            int bSize = sizeMap.get(bParent);
            if (aSize >= bSize) {
                // 让b接到a的下面
                parentMap.put(bParent, aParent);
                sizeMap.put(aParent, aSize + bSize);
                sizeMap.remove(bParent);
            } else {
                // 让a接到b的下面
                parentMap.put(aParent, bParent);
                sizeMap.put(bParent, aSize + bSize);
                sizeMap.remove(aParent);
            }
        }
    }

    private MyNode<V> findParent(MyNode<V> cur) {
        Stack<MyNode<V>> path = new Stack<>();
        // 获取最高的那个祖先
        while (cur != parentMap.get(cur)) {
            path.push(cur);
            cur = parentMap.get(cur);
        }
        // 沿途节点全部优化
        while (!parentMap.isEmpty()) {
            parentMap.put(path.pop(), cur);
        }
        return cur;
    }

}

class MyNode<V> {
    V value;

    public MyNode(V v) {
        value = v;
    }
}